powergenome/util.py

```diff
@@ -59,9 +59,112 @@ def load_settings(path: Union[str, Path]) -> dict:
 
     if settings.get("input_folder"):
         settings["input_folder"] = path.parent / settings["input_folder"]
+
+    settings = apply_all_tag_to_regions(settings)
+
     return fix_param_names(settings)
 
 
+def apply_all_tag_to_regions(settings: dict) -> dict:
+    """Make copies of renewables_clusters dicts with region "all"
+
+    If a renewables clustering object doesn't already existing for a region/technology
+    then make a copy for use. This is helpful with large numbers of regions when
+    the clustering parameters can be applied everywhere.
+
+    Parameters
+    ----------
+    settings : dict
+        All user-specified settings from YAML files
+
+    Returns
+    -------
+    dict
+        Copy of the input settings with renewables_clusters objects for all regions
+
+    Raises
+    ------
+    KeyError
+        The dictionary is missing the tag "region"
+    KeyError
+        The dictionary with region "all" is missing the tag "technology"
+    """
+
+    settings_all = dict()
+    all_regions = settings["model_regions"]
+
+    # Keeps a list of which regions should be modified by "all" (are not specifically tagged)
+    techs_tagged_w_all = []
+    techs_tagged_by_region = dict()
+
+    i = 0
+    to_delete = []
+
+    # These are the keys in settings which will not be used to determine whether 'all' should apply to that region
+    identifier_keys = ["technology", "pref_site", "turbine_type"]
+
+    for d in settings.get("renewables_clusters", {}) or {}:
+
+        if "region" not in d:
+            raise KeyError("Entry missing 'region' tag.")
+
+        reg = d["region"]
+
+        keys = sorted(d.keys())
+        tech = ""
+        for key in keys:
+            if key in identifier_keys:
+                if tech != "":
+                    tech += "_"
+                tech += str(d[key])
+
+        if reg.lower() == "all":
+
+            settings_all[tech] = d
+
+            if "technology" not in d:
+                raise KeyError(f"""Entry for {reg} missing 'technology' tag.""")
+
+            if tech in techs_tagged_w_all:
+                s = f"""
+                Multiple 'all' tags applied to technology {tech}. Only last one will be used.
+                """
+                logger.warning(s)
+
+            else:
+
+                techs_tagged_w_all.append(tech)
+
+            to_delete.append(i)
+
+        else:
+
+            # Update the dict stating that this technology is specified for this region
+            if tech in techs_tagged_by_region:
+                techs_tagged_by_region[tech].append(reg)
+            else:
+                techs_tagged_by_region[tech] = [reg]
+
+        # Keeps track of the "all" tags so that they can be deleted later in the function
+        i += 1
+
+    for i in reversed(to_delete):
+
+        del settings["renewables_clusters"][i]
+
+    for tech in techs_tagged_w_all:
+
+        for reg in all_regions:
+            if reg not in techs_tagged_by_region[tech]:
```

Comment:
@annafjacobson if a tech only has region of "all" then it is never added as a key to techs_tagged_by_region